// @unleash/proxy/dist/metrics-schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientMetricsSchema = exports.countSchema = void 0;
const joi_1 = __importDefault(require("joi"));
exports.countSchema = joi_1.default
    .object()
    .options({ stripUnknown: true })
    .keys({
    yes: joi_1.default.number().min(0).empty('').default(0),
    no: joi_1.default.number().min(0).empty('').default(0),
    variants: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.number().min(0)),
});
exports.clientMetricsSchema = joi_1.default
    .object()
    .options({ stripUnknown: true })
    .keys({
    appName: joi_1.default.string().required(),
    instanceId: joi_1.default.string().required(),
    environment: joi_1.default.string().optional(),
    bucket: joi_1.default
        .object()
        .required()
        .keys({
        start: joi_1.default.date().required(),
        stop: joi_1.default.date().required(),
        toggles: joi_1.default.object().pattern(/.*/, exports.countSchema),
    }),
});
